pymc_experimental/inference/pathfinder/importance_sampling.py

```diff
@@ -0,0 +1,142 @@
+
+    References
+    ----------
+    Elvira, V., Martino, L., Luengo, D., & Bugallo, M. F. (2019). Generalized Multiple Importance Sampling. Statistical Science, 34(1), 129-155. https://doi.org/10.1214/18-STS668
+
+    Yao, Y., Vehtari, A., Simpson, D., & Gelman, A. (2018). Yes, but Did It Work?: Evaluating Variational Inference. arXiv:1802.02538 [Stat]. http://arxiv.org/abs/1802.02538
+
+    Zhang, L., Carpenter, B., Gelman, A., & Vehtari, A. (2022). Pathfinder: Parallel quasi-Newton variational inference. Journal of Machine Learning Research, 23(306), 1-49.
+    """
+
+    num_paths, num_pdraws, N = samples.shape
+
+    if method == "none":
+        logger.warning(
+            "importance sampling is disabled. The samples are returned as is which may include samples from failed paths with non-finite logP or logQ values. It is recommended to use importance_sampling='psis' for better stability."
+        )
+        return samples
+    else:
+        samples = samples.reshape(-1, N)
+        logP = logP.ravel()
+        logQ = logQ.ravel()
+
+        # adjust log densities
+        log_I = np.log(num_paths)
+        logP -= log_I
+        logQ -= log_I
+        logiw = logP - logQ
+
+        if method == "psis":
+            replace = False
+            logiw, pareto_k = PSIS()(logiw)
+        elif method == "psir":
+            replace = True
+            logiw, pareto_k = PSIS()(logiw)
+        elif method == "identity":
+            replace = False
+            logiw = logiw
+            pareto_k = None
+        else:
+            raise ValueError(f"Invalid importance sampling method: {method}")
+
+    # NOTE: Pareto k is normally bad for Pathfinder even when the posterior is close to the NUTS posterior or closer to NUTS than ADVI.
+    # Pareto k may not be a good diagnostic for Pathfinder.
+    if pareto_k is not None:
+        pareto_k = pareto_k.eval()
+        if pareto_k < 0.5:
+            pass
+        elif 0.5 <= pareto_k < 0.70:
+            logger.info(
+                f"Pareto k value ({pareto_k:.2f}) is between 0.5 and 0.7 which indicates an imperfect approximation however still useful."
+            )
+            logger.info("Consider increasing ftol, gtol, maxcor or num_paths.")
+        elif pareto_k >= 0.7:
+            logger.info(
+                f"Pareto k value ({pareto_k:.2f}) exceeds 0.7 which indicates a bad approximation."
+            )
+            logger.info(
+                "Consider increasing ftol, gtol, maxcor, num_paths or reparametrising the model."
+            )
+        else:
+            logger.info(
+                f"Received an invalid Pareto k value of {pareto_k:.2f} which indicates the model is seriously flawed."
+            )
+            logger.info(
+                "Consider reparametrising the model all together or ensure the input data are correct."
+            )
+
+        logger.warning(f"Pareto k value: {pareto_k:.2f}")
+
+    p = pt.exp(logiw - pt.logsumexp(logiw)).eval()
```

Comment:
You shouldn't be using pytensor if you're evaling immediately. Just use numpy/scipy

Reply:
It depends on whether you're woring with pytensor graphs or numpy arrays